[lldb] Correctly restore the cursor column after resizing the statusline

JDevlieghere opened the pull request:
This PR ensures we correctly restore the cursor column after resizing the statusline. To ensure we have space for the statusline, we have to emit a newline to move up everything on screen. The newline causes the cursor to move to the start of the next line, which needs to be undone.

Normally, we would use escape codes to save & restore the cursor position, but that doesn't work here, as the cursor position may have (purposely) changed. Instead, we move the cursor up one line using an escape code, but we weren't restoring the column.

Interestingly, Editline was able to recover from this issue through the LineInfo struct which contains the buffer and the cursor location, which allows us to compute the column. This PR addresses the bug by relying on the active IOHandler to report its cursor position and if available, use that to restore the cursor column position.

Fixes #134064

DavidSpickett commented on lldb/source/Host/common/Editline.cpp:
```diff
@@ -122,22 +122,21 @@ static int GetOperation(HistoryOperation op) {
+  case HistoryOperation::Newer:
+    return H_PREV;
+  case HistoryOperation::Newest:
+    return H_FIRST;
```
Some stray clang-format changes here and below.

DavidSpickett commented on lldb/source/Host/common/Editline.cpp:
```diff
@@ -398,6 +397,20 @@ int Editline::GetLineIndexForLocation(CursorLocation location, int cursor_row) {
+    return {};
+
+  const size_t editline_cursor_col =
+      (int)((info->cursor - info->buffer) + GetPromptWidth()) + 1;
```
You're going to static cast this later, so could you make the type here unsigned to begin with?

JDevlieghere replied:
The static cast is to appease the printf-format specifier. Storing the position as a `size_t` seems more canonical, but I also can't imagine a terminal that exceeds `UINT32_MAX`

DavidSpickett commented on lldb/source/Core/Statusline.cpp:
```diff
@@ -102,20 +103,36 @@ void Statusline::UpdateScrollWindow(ScrollWindowMode mode) {
   const unsigned scroll_height =
       (mode == DisableStatusline) ? m_terminal_height : m_terminal_height - 1;
 
+  CursorPosition cursor_position = m_debugger.GetIOHandlerCursorPosition();
```
Move this into the if, I know it has to be done before the newline, but it can still be closer to point of use.

DavidSpickett reviewed (commented):
So if I understand correctly, the newline acts like `\n` and `\r`. `\n` moves the cursor down one line and `\r` moves it to the start of that line.

You're undoing the `\n` by moving up one row, and the `\r` by restoring the column position.

DavidSpickett commented:
And how exactly does this fix https://github.com/llvm/llvm-project/issues/134064?

As in, what steps are happening with this PR compared to before.

JDevlieghere commented:
> And how exactly does this fix #134064?

https://github.com/llvm/llvm-project/issues/134064#issuecomment-2923315404 has a visualization of the problem, which you've summarized in https://github.com/llvm/llvm-project/pull/145823#pullrequestreview-2961296028. 

> As in, what steps are happening with this PR compared to before.

This PR uses the active IOHandler (i.e. Editline) to get the cursor position before we emit the newline and then uses an ANSI escape code to move the cursor back to its original column. I was curious why, depending on the timing, Editline was able to recover from this and this is how Editline does it.

JDevlieghere committed:
[lldb] Correctly restore the cursor column after resizing the statusline

This PR ensures we correctly restore the cursor column after resizing
the statusline. To ensure we have space for the statusline, we have to
emit a newline to move up everything on screen. The newline causes the
cursor to move to the start of the next line, which needs to be undone.

Normally, we would use escape codes to save & restore the cursor
position, but that doesn't work here, as the cursor position may have
(purposely) changed. Instead, we move the cursor up one line using an
escape code, but we weren't restoring the column.

Interestingly, Editline was able to recover from this issue through the
LineInfo struct which contains the buffer and the cursor location, which
allows us to compute the column. This PR addresses the bug by relying on
the active IOHandler to report its cursor position and if available, use
that to restore the cursor column position.

Fixes #134064

DavidSpickett commented:
Ok I get what we're doing now but...

> This PR uses the active IOHandler (i.e. Editline) to get the cursor position before we emit the newline and then uses an ANSI escape code to move the cursor back to its original column. I was curious why, depending on the timing, Editline was able to recover from this and this is how Editline does it.

This PR uses the active IOHanlder (which is Editline), but you're saying that prior to this PR, Editline was sometimes able to recover. So we were using Editline all along? Confused as to what is Editline the library and what is our Editline code.

Do you mean that it's now all done in the same place and that keeps the timing correct?

DavidSpickett reviewed (approved):
I am confused which bits are handled by editline the library and which are handled by our code that interacts with editline, but I always have been so that shouldn't block this :)

LGTM

JDevlieghere commented:
> This PR uses the active IOHanlder (which is Editline), but you're saying that prior to this PR, Editline was sometimes able to recover. So we were using Editline all along? Confused as to what is Editline the library and what is our Editline code.
> I am confused which bits are handled by editline the library and which are handled by our code that interacts with editline, but I always have been so that shouldn't block this :)

I didn't quite get what was causing the confusion so thanks for clarifying. 

1. The bug is always there: when reducing the scroll window (on launch or when resizing the terminal) the statusline always leaves the cursor at the wrong column.
2. The bug isn't always observable: when Editline (re)draws the prompt after we left the cursor in the wrong spot it looks like things behave correctly, even though they don't. The bug is there but it's obscured. 
3. When you hit the bug, you can recover from it by typing something. This is what I meant when I said that "Editline is able to recover from this". 

The challenge is that there is no escape code to move the cursor to its correct location, at least not without knowing what that location is. I was intrigued by the fact that Editline seemingly was able to do that anyway, so I wanted to understand how (2) and especially (3).  The former could have been explained by just redrawing the prompt, but I noticed that if your cursor was't right at the end of the prompt, (3) would still do the right thing. 

The answer was that `Editline::MoveCursor` is using the `LineInfo` struct to get the index of the cursor and the length of the buffer, which allows you to compute the position of the cursor on the current line. In other words, when Editline was redrawing the cursor, it was calling `Editline::MoveCursor` to move the cursor to the correct position.

One solution could have been to indirectly rely on this behavior and ask our Editline wrapper to redraw the prompt or something after we change the scroll window. That's essentially was @labath suggested in https://github.com/llvm/llvm-project/issues/134064#issuecomment-2929542360. However that's overkill: instead of using a hammer and redraw the whole thing, we can use the same information (the `LineInfo` struct in Editline) to figure out the correct column position. That's what this patch does: it queried the IOHandler (Editline) for the cursor column position before we much with it, and then uses that to move the cursor to that exact location. 

In other words, we're using the same information that Editline uses to move the cursor to the correct position when drawign the statusline. But we're doing it without having Editline do that for us.

DavidSpickett commented:
> When you hit the bug, you can recover from it by typing something. This is what I meant when I said that "Editline is able to recover from this".

Understood. Editline does not avoid the situation, it provides a way out of it.

JDevlieghere commented:
> That's true, but it kind of relies on editline not moving the cursor between the time we ask for the location and when we do the statusline dance. In that sense, the hammer approach would be more robust as it would presumably happen while holding the editline mutex. I guess that would be fine in most cases because changing the setting is usually triggered by a CLI command, and that should complete before we start processing the next command... Though, if that's the case, then I'm confused why isn't this working already, and if we couldn't just reorder a couple of calls to make sure the status line is drawn before the prompt.

Holding the  editline mutex is a compelling argument for the hammer approach. For completeness, the other reasons I didn't go with this approach were that we already call back from Editline to redraw the statusline after it has done something to clear the screen so you'd get a dance where editline asks the statusline to redraw which asks editline to redraw the prompt. 

I'll take a look and see what it would take to create a function in Editline that can "redraw the prompt" safely. The `MoveCursor` function takes an argument that tells it where the cursor is (i.e. `EditingCursor`, `EditingPrompt`, `BlockEnd`) based on where it's called from. I'll need to make sure this behaves correctly when called "asynchronously". 

> Going back to my list of bad ideas, I think I now know why the first one doesn't work (because we don't know whether `\n'` will scroll the window -- it will only do that if we're at the last line), but here's bad idea number four: What if we just _always_ scroll the window -- instead of `\n` we do a `Scroll Up(1)`? Could we then do the dance with saving the cursor location? I guess it would mean that we lose one line of visible output in the case where it would not be necessary to scroll (cursor not at last line), but maybe that's not such a bad thing?

I had considered this but I quickly discarded it after trying it because it looks like very much like a bug. When you interactive resize the terminal we'll get a bunch of `SIGWINCH`s (which is desirable because otherwise text might have wrapped and we don't get a chance to clear it, which is another bug...) so your text would scroll up every time the number of terminal rows changes, which can be a lot.

JDevlieghere commented:
Alternative approach: https://github.com/llvm/llvm-project/pull/146132

labath commented:
Okay, I think I see the source of confusion. I was expecting the terminal to keep the status line at the bottom of the screen after a resize (just like it does with regular text), but my terminal (XFCE) at least simply discards it when decreasing the window size. It makes kind of sense, I guess, but it means my idea won't work. I think the other patch is a good solution to this problem.
